src/Features/Core/Portable/Diagnostics/HostAnalyzerManager.cs

```diff
@@ -95,65 +120,60 @@ public DiagnosticAnalyzerInfoCache(Lazy<ImmutableArray<HostDiagnosticAnalyzerPac
-        {
-            return reference.Id;
-        }
+            => reference.Id;
 
         /// <summary>
         /// It returns a map with <see cref="AnalyzerReference.Id"/> as key and <see cref="AnalyzerReference"/> as value
         /// </summary>
-        public ImmutableDictionary<object, AnalyzerReference> CreateAnalyzerReferencesMap(Project? project = null)
-        {
-            if (project == null)
-            {
-                return _hostAnalyzerReferencesMap.Value;
-            }
-
-            return _hostAnalyzerReferencesMap.Value.AddRange(CreateProjectAnalyzerReferencesMap(project));
-        }
+        public ImmutableDictionary<object, AnalyzerReference> GetHostAnalyzerReferencesMap()
+            => _hostAnalyzerReferencesMap.Value;
 
         /// <summary>
-        /// Return <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> of given <paramref name="analyzer"/>.
+        /// Returns <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> of given <paramref name="analyzer"/>.
         /// </summary>
         public ImmutableArray<DiagnosticDescriptor> GetDiagnosticDescriptors(DiagnosticAnalyzer analyzer)
-        {
-            return (ImmutableArray<DiagnosticDescriptor>)_descriptorCache.GetValue(analyzer, a => GetDiagnosticDescriptorsCore(a));
-        }
+            => GetOrCreateDescriptorsInfo(analyzer).SupportedDescriptors;
 
-        private ImmutableArray<DiagnosticDescriptor> GetDiagnosticDescriptorsCore(DiagnosticAnalyzer analyzer)
+        /// <summary>
+        /// Determine whether collection of telemetry is allowed for given <paramref name="analyzer"/>.
+        /// </summary>
+        public bool IsTelemetryCollectionAllowed(DiagnosticAnalyzer analyzer)
+            => GetOrCreateDescriptorsInfo(analyzer).TelemetryAllowed;
+
+        private DiagnosticDescriptorsInfo GetOrCreateDescriptorsInfo(DiagnosticAnalyzer analyzer)
+            => _descriptorsInfo.GetValue(analyzer, CalculateDescriptorsInfo);
+
+        private DiagnosticDescriptorsInfo CalculateDescriptorsInfo(DiagnosticAnalyzer analyzer)
         {
             ImmutableArray<DiagnosticDescriptor> descriptors;
             try
             {
                 // SupportedDiagnostics is user code and can throw an exception.
-                descriptors = analyzer.SupportedDiagnostics;
-                if (descriptors.IsDefault)
-                {
-                    descriptors = ImmutableArray<DiagnosticDescriptor>.Empty;
-                }
+                descriptors = analyzer.SupportedDiagnostics.NullToEmpty();
             }
-            catch (Exception ex)
+            catch
             {
-                AnalyzerHelper.OnAnalyzerExceptionForSupportedDiagnostics(analyzer, ex, _hostDiagnosticUpdateSource);
+                // No need to report the exception to the user.
+                // Eventually, when the analyzer runs the compiler analyzer driver will report a diagnostic.
```

Comment:
One case I had forgotten where this might not be true is for IDE-only analyzers, such as the ones deriving from DocumentDiagnosticAnalyzer and ProejctDiagnosticAnalyzer. However, I think we can assume this are internal ones, which should ideally never throw in SupportedDiagnostics.